server/gitrest/packages/gitrest-base/src/utils/gitWholeSummaryManager.ts

```diff
@@ -101,45 +141,83 @@ function getSummaryObjectFromWholeSummaryTreeEntry(entry: WholeSummaryTreeEntry)
     throw new NetworkError(400, `Unknown entry type: ${entry.type}`);
 }
 
-interface IFullGitTree {
-    tree: ITree;
-    blobs: Record<string, IBlob>;
-}
-function containsFullGitTree(gitTree: ITree): boolean {
-    return gitTree.tree.length === 1
-        && gitTree.tree[0].type === "blob"
-        && gitTree.tree[0].path.endsWith(fullTreePath);
-}
+/**
+ * Package a Git tree into a Full Git Tree object by (optionally) parsing and unpacking
+ * inner full git tree blobs and (optionally) retrieving all referenced blobs from storage.
+ * 
+ * @param gitTree - Git tree object containing tree entries
+ * @param repoManager - Repository manager to use for retrieving referenced blobs
+ * @param parseInnerFullGitTrees - Whether to parse and unpack inner full git tree blobs
+ * @param retrieveBlobs - Whether to retrieve blobs (other than full git trees) from storage
+ * @param depth - internally tracks recursion depth for potential future logging or protection
+ */
 async function buildFullGitTreeFromGitTree(
     gitTree: ITree,
     repoManager: IRepositoryManager,
+    parseInnerFullGitTrees = true,
+    retrieveBlobs = true,
+    depth = 0,
 ): Promise<IFullGitTree> {
+    let parsedFullTreeBlobs = false;
     const blobPs: Promise<IBlob>[] = [];
-    gitTree.tree.forEach((treeEntry) => {
+    const treeEntries: ITreeEntry[] = [];
+    for (const treeEntry of gitTree.tree) {
         if (treeEntry.type === "blob") {
-            blobPs.push(repoManager.getBlob(treeEntry.sha));
+            if (treeEntry.path.endsWith(fullTreePath) && parseInnerFullGitTrees) {
+                parsedFullTreeBlobs = true;
+                const fullTreeBlob = await repoManager.getBlob(treeEntry.sha);
+                const fullTree = JSON.parse(
+                    fullTreeBlob.encoding === "base64"
+                        // Convert base64 to utf-8 for JSON parsing
+                        ? Buffer.from(fullTreeBlob.content, fullTreeBlob.encoding).toString("utf-8")
+                        : fullTreeBlob.content,
+                ) as IFullGitTree;
+                const builtFullGitTree = await buildFullGitTreeFromGitTree(
+                    fullTree.tree,
+                    repoManager,
+                    true, /* parseInnerFullGitTrees */
+                    // All blobs associated with full git tree are included in the full git tree blob, and
+                    // will not exists in storage individually.
+                    false, /* retrieveBlobs */
+                    depth + 1,
+                );
+                const baseTreeEntryPath = treeEntry.path.replace(fullTreePath, "");
+                treeEntries.push(...builtFullGitTree.tree.tree.map((fullTreeEntry => ({
+                    ...fullTreeEntry,
+                    path: `${baseTreeEntryPath}${fullTreeEntry.path}`,
+                }))));
+                const fullTreeBlobs = {
+                    ...fullTree.blobs,
+                    ...builtFullGitTree.blobs,
```

Comment:
Just to make sure that my mental model is correct:
1. `fullTree.blobs` are the top level blobs, included at the full tree (encoded as a single blob), depth === 0
2. `builtFullGitTree.blobs` would include the other possible blobs at depths === 1+. Those would be obtaining by using the top level tree (the blobs were already taken care of in step 1 above) to recursively obtain those blobs. 

Is that correct? The part that confused me a bit is what the tree would look like when in the recursion loop. This comment said:
````
// All blobs associated with full git tree are included in the full git tree blob, and
```
Does that mean that the `fullTree.tree` that is passed to `buildFullGitTreeFromGitTree` will also have the `fullTreePath` encoded in its sha?

Reply:
> Just to make sure that my mental model is correct:
> - `fullTree.blobs` are the top level blobs, included at the full tree (encoded as a single blob), depth === 0
> - `builtFullGitTree.blobs` would include the other possible blobs at depths === 1+. Those would be obtaining by using the top level tree (the blobs were already taken care of in step 1 above) to recursively obtain those blobs.

This is almost correct. The part that's missing is when `fullTree.blobs` contains a fullTree single blob in itself, in which case we need to parse out and unpack that internal fullTree single blob, which will give us a new list of blobs that were hidden within the internal fullTreeSingle blob. Only the internal fullTree single blob is contained in the fullTree.blobs, not the internal fullTree's internal blobs. We pass `false, /* retrieveBlobs */` into the recursive `buildFullGitTreeFromGitTree` calls because we know that any blobs not contained in an internal fullGitTree blob are contained in the current fullGitTree blob.

I might not be explaining that well, but hopefully it kinda makes sense.

Reply:
> Does that mean that the `fullTree.tree` that is passed to `buildFullGitTreeFromGitTree` will also have the `fullTreePath` encoded in its sha?

This is a different question entirely, because there are a lot of reasons that the `fullGitTree-Blob` sha would be different from the `fullGitTree` original tree sha. A lot of the complexity of this PR is due to this being the case, and there's not really a way to workaround it (that I could think of). That's OK though, because "unpacking" and writing the "packed" git tree back into Git (in-memory) as a normal Git Tree results in the same sha as the original git tree that was "packed".